pagefaultgames/pokerogue
# [Item/Balance] Overhaul Lapsing Modifiers

## What are the changes the user will see?


- General Changes
  - Both lures (all tiers) and battle items have their battle counters adjusted such that they begin with solid green outline and, as they lapse over battles, turn to dark red.
  - Both items, when "refreshed", play a sound cue (`se/restore`) to indicate that they've changed. 
  - Both items now have updated English entries to describe their new behavior.
- Changes to Lures
  - Lures now mirror the behavior of battle items in that multiple lures of the same type can no longer be held at once and getting another will only "refresh" the battle count of the existing same-type lure; you can still have every tier of lure at the same time with stacking applications to double battle chance. 
  - To maintain the consequential behavior of having multiple lures at a given time, lures now provide a 4x chance to provide a double battle. As a result, there are new double battle probabilities for owning each lure:
    - One unique lure -> 50% double battle chance, 12.5% double boss battle chance
    - Two unique lures -> 100% double battle chance, 50% double boss battle chance
    - Three unique lures -> 100% double battle chance, 100% double boss battle chance

## Why am I making these changes?




This PR rounds out the battle item changes made in #2699 by providing up-to-date descriptions and provides better visual feedback for what is essentially new behavior to the game entirely ("refreshing" lapsing modifiers instead of stacking multiple on top of one another). 

## What are the changes from a developer perspective?



Originally, lures were left alone due to being out of scope of the main intent of #2699 and nerfing battle items in general; as such, the battle count reset logic in `add()` had to be duplicated for both `Dire Hit` and the rest of the battle items. By including lures, the code can now make use of polymorphism in the extended class, `LapsingPersistentModifier`, which makes all classes involved more clean overall. In addition, a `maxBattles` field was added in order to keep track of what value to refresh back to aid in accurate linear interpolation

A helper function `hslToHex()` was added to `Utils` to essentially allow for on-the-fly hex code generation with varying hue, saturation, and lightness. In this specific case, linear interpolation is used on hue (formula below) while keeping saturation and lightness constant to achieve the gradual hue shifting on the items' battle count text:

$$ \text{hue} = m \cdot \frac{\text{battleCount}}{\text{maxBattles}} + n $$

where $m + n$ is the ending value and $n$ is the starting value. Note that these can be tweaked and adjusted as needed; none of the colors are hard-coded.

Unit tests were added for lures and updated for battle items to reflect the changes (note that visual changes are not tested). In addition to this, an `item` option was added to the test boilerplate script.

### Screenshots/Videos



#### Example - `Dire Hit` Lapsing
| 5 | 4 | 3 | 2 | 1 |
| --- | --- | --- | --- | --- |
| <img width="81" alt="5" src="https://github.com/user-attachments/assets/683dac29-d321-4400-b270-128770ddb0ec"> | <img width="81" alt="4" src="https://github.com/user-attachments/assets/78755394-4232-4587-88dc-2c186aba996c"> | <img width="81" alt="3" src="https://github.com/user-attachments/assets/b0b4cff5-8bd7-401f-accd-80ae1a34bf54"> | <img width="81" alt="2" src="https://github.com/user-attachments/assets/39b62183-7ba0-4242-8cf1-31ff0e81f3b7"> | <img width="81" alt="1" src="https://github.com/user-attachments/assets/dc8180af-7ec1-4183-a3b2-669deee5e269"> |

#### Example - Save Slot

<img width="648" alt="saveSlot" src="https://github.com/user-attachments/assets/94a00b31-d0a6-4108-9955-c9662fa29bf6">

#### Examples - In-Battle

| Fresh Lapsing Modifiers | Lapsing Modifiers after 4 Battles |
| --- | --- |
| <img width="662" alt="earlyBattle" src="https://github.com/user-attachments/assets/f7d6cc30-6cb0-49a3-b0d9-9d2b45849502"> | <img width="662" alt="lateBattle" src="https://github.com/user-attachments/assets/6e82dd2e-b7cb-4952-b9f0-e26113ba3579"> |

https://github.com/user-attachments/assets/7af1e819-1fdd-441d-8366-81525d9d425f

## How to test the changes?





The most fun way to test the new behavior and observe the new visual effect is by using the following overrides and starting a new Classic run:
```ts
const overrides = {
  ITEM_REWARD_OVERRIDE: [
    { name: "TEMP_STAT_STAGE_BOOSTER", type: Stat.ATK },
    { name: "DIRE_HIT" },
    { name: "SUPER_LURE" }
  ],
  STARTING_MODIFIER_OVERRIDE: [
    { name: "TEMP_STAT_STAGE_BOOSTER", type: Stat.ATK },
    { name: "TEMP_STAT_STAGE_BOOSTER", type: Stat.DEF },
    { name: "TEMP_STAT_STAGE_BOOSTER", type: Stat.SPATK },
    { name: "TEMP_STAT_STAGE_BOOSTER", type: Stat.SPDEF },
    { name: "TEMP_STAT_STAGE_BOOSTER", type: Stat.SPD },
    { name: "TEMP_STAT_STAGE_BOOSTER", type: Stat.ACC },
    { name: "DIRE_HIT" },
    { name: "LURE" },
    { name: "SUPER_LURE" },
    { name: "MAX_LURE" }
  ],
  STARTING_LEVEL_OVERRIDE: 100
} satisfies Partial<InstanceType<typeof DefaultOverrides>>;
```

The item rewards are used to refresh any items of the same type if applicable.

## Checklist
